### ISSUE TITLE
VSCode setting should be overriden by discovered toml

### ISSUE THREAD

**Issue by celynw** [astral-sh/ruff-vscode#425]
Currently, extension settings override those which would be picked up in a `pyproject.toml`/`ruff.toml`.
This was highlighted in #3:

> Yeah, that's correct -- the `pyproject.toml` in the project path _doesn't_ override the settings from the VS Code extension. Which is not ideal, for the reasons you described.
> 
> The best way to solve this _today_ is to use a user-specific `pyproject.toml` rather than going through the VS Code extension. For example, if you're on macOS, you can create a `pyproject.toml` at `~/Library/Application\ Support/ruff/pyproject.toml`. Then, when linting, if Ruff can't find a `pyproject.toml` in the current file path, it will use _that_ as the default settings.

This is still the case (in my testing), but I'm surprised not to see another issue covering this (happy to be corrected!)
I think there should at least be an issue to track this one.

## Environment

- Ruff extension version `v2024.16.0` (`ruff==0.3.1`)
- VSCode version `1.87.2`
- OS: `Ubuntu 22.04.4`

Tested simply with `line-length = 120` in `ruff.toml`, and `--line-length=120` in VSCode setting `ruff.lint.args`

**Comment by MichaReiser** [comment]
I think this needs some design work on the intended behavior (CC: @snowsignal). To me, the VS code settings taking precedence over the project settings seem reasonable, or at least I could see where the opposite behavior is the intended behavior.  It also opens the question if this should apply to all options or only some (e.g. what if I configure an additional configuration?) But I agree that what you're asking for makes sense, too; ultimately, the experience between IDE and CLI should be consistent, and that is only guaranteed by giving the project. to a higher priority.

**Comment by celynw** [comment]
My use case is that I have synchronised VSCode settings everywhere. I'd like to be able to open different projects which each contain a `pyproject.toml` file. I either have to keep my 'master' synchronised Ruff VSCode settings empty, or I have to create a workspace settings file (under `.vscode/settings.json`) explicitly setting those settings to be empty, overriding my 'master' synced settings and allowing the toml file to do something.

But yes, I do see why one might expect the VSCode setting to take precedence instead!

**Comment by snowsignal** [comment]
I'd support having a flag in the extension that lets discovered configuration take precedence. We'll look into solving this as we work on configuration for the upcoming VSCode extension based on our new Rust-based LSP.

**Comment by dhruvmanila** [comment]
This is possible now with the new [`ruff.configurationPreference`](https://docs.astral.sh/ruff/editors/settings/#configurationpreference) editor setting.

**Comment by beauxq** [comment]
> This is possible now with the new [`ruff.configurationPreference`](https://docs.astral.sh/ruff/editors/settings/#configurationpreference) editor setting.

Thanks for this.

I would recommend changing it to be the default.
It was pretty unexpected for it not to work this way, and it's different from other tools like pyright. If pyright finds a text configuration file (either `[tool.pyright]` in `pyproject.toml` or `pyrightconfig.json`), it uses that and ignores what is in VSCode settings, and this was never surprising to me.

With this Ruff extension, I was confused and surprised and had trouble figuring out what was going on.

**Comment by dhruvmanila** [comment]
> I would recommend changing it to be the default.
> It was pretty unexpected for it not to work this way, and it's different from other tools like pyright. If pyright finds a text configuration file (either `[tool.pyright]` in `pyproject.toml` or `pyrightconfig.json`), it uses that and ignores what is in VSCode settings, and this was never surprising to me.

I've opened https://github.com/astral-sh/ruff/issues/13067 to track that.

**Comment by VictorGillioz** [comment]
Thanks for implementing the [ruff.configurationPreference](https://docs.astral.sh/ruff/editors/settings/#configurationpreference) setting. I wanted to raise one behavior that I found confusing:

If some ruff settings are not set in the `.toml` file, the extension will use preferences from the editor, not the default values. I assume this is what "takes priority" mean?

This results in undesirable behaviors, including inconsistencies in linting and formatting. For example, if line length is not defined in the `.toml` file but defined in `settings.json`, running `ruff format` in the terminal will apply the default of 88 while vscode's format will apply the `settings.json` length.

Maybe there is an easy workaround I missed? If not, would it be possible to have a "filesystemOnly" option that would use the editor's config only when no `.toml` is present? Alternatively, could we have "hard" vs "soft" priorities for both "editorFirst" and "filesystemFirst"?

Thanks in advance

**Comment by dhruvmanila** [comment]
> If some ruff settings are not set in the `.toml` file, the extension will use preferences from the editor, not the default values. I assume this is what "takes priority" mean?

Yes, that's correct. The settings are merged and the linked setting specifies which one should be preferred.

> This results in undesirable behaviors, including inconsistencies in linting and formatting. For example, if line length is not defined in the `.toml` file but defined in `settings.json`, running `ruff format` in the terminal will apply the default of 88 while vscode's format will apply the `settings.json` length.
> 
> Maybe there is an easy workaround I missed? If not, would it be possible to have a "filesystemOnly" option that would use the editor's preferences only when no `.toml` is present? Alternatively, could we have "hard" vs "soft" priorities for both "editorFirst" and "filesystemFirst"?

I might be naive here but just curious to understand why have a value in the editor setting when it _shouldn't_ be used? Regardless, it might make sense to have a "filesystemOnly" value similar to "editorOnly".

**Comment by VictorGillioz** [comment]
> I might be naive here but just curious to understand why have a value in the editor setting when it shouldn't be used?

I find it practical to have preferences set in VS Code's User Settings for early/quick/informal code development where a .toml does not exist.

However, it means these preferences will be used over default values in projects where a .toml file exists but does not define them, resulting in the inconsistency between terminal and editor execution.

Is there another approach you would recommend?

**Comment by dhruvmanila** [comment]
> However, it means these preferences will be used over default values in projects where a .toml file exists but does not define them, resulting in the inconsistency between terminal and editor execution.
> 
> Is there another approach you would recommend?

I don't think I have any other approach in mind which can help in your use-case.

I think it makes sense to add `filesystemOnly` and the implementation wouldn't be too difficult. I'll open an issue for this specifically (https://github.com/astral-sh/ruff/issues/16209)

**Comment by dhruvmanila** [comment]
Closing this as it's implemented.
